File: Tar/Java&Plat/Demos/Leetcode/WklyCont/src/main/java/com/essenbazar/_240/Solution1856.java
package com.essenbazar._240;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/*

The min-product of an array is equal to the minimum value in the array multiplied by the array's sum.

    For example, the array [3,2,5] (minimum value is 2)
        has a min-product of 2 * (3+2+5) = 2 * 10 = 20.

Given an array of integers nums, return the maximum min-product of any non-empty
    subarray of nums.

    Since the answer may be large, return it modulo 109 + 7.

Note that the min-product should be maximized before performing the modulo operation.
    Testcases are generated such that the maximum min-product without modulo will
    fit in a 64-bit signed integer.

A subarray is a contiguous part of an array.

Input: nums = [1,2,3,2]
Output: 14
Explanation: The maximum min-product is achieved with the subarray [2,3,2] (minimum value is 2).
2 * (2+3+2) = 2 * 7 = 14.

 */

class Solution1856 {

	public static int maxSumMinProduct(int[] a) {
		long MOD = (int) (1e9 + 7);
		int n = a.length;
		if (n == 1) {
			return (int) (((long) a[0] * (long) a[0]) % MOD);
		}
		long res = 0L;
		long[] preS = new long[n + 1];
		preS[0] = 0L;
		int ii = 1;
		for (int num : a) {
			preS[ii] = preS[ii - 1] + num;
			ii++;
		}
		Stack<List<Integer>> stack = new Stack<>();
		for (int i = 0; i < a.length; ++i) {
			int newStart = i;
			while (!stack.isEmpty() && stack.peek().get(1) > a[i]) {
				List<Integer> check_ = stack.pop();
				int start = check_.get(0);
				int value = check_.get(1);
				long total = preS[i] - preS[start];
				res = Math.max(res, value * total);
				newStart = start;
			}
			List<Integer> arr_ = new ArrayList<>();
			arr_.add(newStart);
			arr_.add(a[i]);
			stack.add(arr_);
		}
		for (List<Integer> check : stack) {
			long total = preS[a.length] - preS[check.get(0)];
			res = Math.max(res, check.get(1) * total);
		}
		return (int) (res % MOD);
	}
}

File: Tar/Python/Leetcode/WeeklyContest/240/Solution1856.py
'''
The min-product of an array is equal to the minimum value in the array multiplied by the array's sum.

    For example, the array [3,2,5] (minimum value is 2) has a min-product of 2 * (3+2+5) = 2 * 10 = 20.

Given an array of integers nums, return the maximum min-product of any non-empty subarray of nums. 
Since the answer may be large, return it modulo 10^9 + 7.

Note that the min-product should be maximized before performing the modulo operation. 
Testcases are generated such that the maximum min-product without modulo will fit in a 64-bit 
signed integer.

A subarray is a contiguous part of an array.

Input: nums = [1,2,3,2]
Output: 14
Explanation: The maximum min-product is achieved with the subarray [2,3,2] (minimum value is 2).
2 * (2+3+2) = 2 * 7 = 14.

'''


class Solution:

    def maxSumMinProduct(self, nums):
        res = 0
        stack = []  # will be monotonic increasing order
        prefixArr = [0]

        for n in nums:
            prefixArr.append(prefixArr[-1] + n)

        for (i, n) in enumerate(nums):
            newStart = i

            # if our stack is non empty and 'stack's top's value' is greater than n

            while stack and stack[-1][1] > n:
                (start, value) = stack.pop()

                # since this value is now being popped, what we are saying is that
                # we just determined what is the max sub-array it could possibly have been a part of
                # so now we're end up computing what the minimum product it would have been

                total = prefixArr[i] - prefixArr[start]

                res = max(res, value * total)

                # and last but not the least, what we're gonna do since we popped a value
                # the end value we will ultimately be adding to our stack

                newStart = start

            stack.append((newStart, n))

            # by the end of the while loop we'll have added every single value to the stack
            # but we won't necessarily have popped every value from the stack

        for (start, val) in stack:
            total = prefixArr[len(nums)] - prefix[start]
            res = max(res, val * total)

        return res % (10 ** 9 + 7)
